acra-censor/handlers/handlers_util.go

```diff
@@ -78,3 +68,189 @@ func RedactSQLQuery(sql string) (string, error) {
+			if queryNodeColName, ok := queryNodes[index].(*sqlparser.ColName); ok && queryNodeColName != nil {
+				if strings.EqualFold(patternNodeColName.Name.String(), ColumnConfigPlaceholderReplacer) {
+					matchDetected = true
+				} else {
+					return false
+				}
+			}
+		}
+	}
+	return matchDetected
+}
+
+// handleSelectWherePattern handles SELECT a, b from t %%WHERE%% pattern
+func handleSelectWherePattern(queryNodes, patternNodes []sqlparser.SQLNode) bool {
+	patternWhereDetected := false
+	queryWhereDetected := false
+	for index, patternNode := range patternNodes {
+		if index >= len(queryNodes) {
+			return false
+		}
+		if index == 0 || reflect.DeepEqual(queryNodes[index], patternNode) {
+			continue
+		}
+		if patternWhereNode, ok := patternNode.(*sqlparser.Where); ok && patternWhereNode != nil && strings.EqualFold(sqlparser.String(patternWhereNode.Expr), WhereConfigPlaceholderReplacerPart2) {
+			patternWhereDetected = true
+		}
+		if queryWhereNode, ok := queryNodes[index].(*sqlparser.Where); ok && queryWhereNode != nil {
+			queryWhereDetected = true
+		}
+		if queryWhereDetected && patternWhereDetected {
+			return true
+		}
+		return false
+	}
+	//this is a case when pattern == query
+	return true
+}
+
+// handle SELECT a, b FROM t1 WHERE userID=%%VALUE%% pattern
+func handleValuePattern(queryNodes, patternNodes []sqlparser.SQLNode) bool {
+	matchDetected := false
+	if len(patternNodes) != len(queryNodes) {
+		return false
+	}
+	for index, patternNode := range patternNodes {
+		if index == 0 || reflect.DeepEqual(patternNode, queryNodes[index]) {
+			continue
+		}
+		if patternNodeComparison, ok := patternNode.(*sqlparser.ComparisonExpr); ok && patternNodeComparison != nil {
+			if queryNodeComparison, ok := queryNodes[index].(*sqlparser.ComparisonExpr); ok && queryNodeComparison != nil {
+				if reflect.DeepEqual(queryNodeComparison.Left, patternNodeComparison.Left) {
+					if strings.EqualFold(sqlparser.String(patternNodeComparison.Right), ValueConfigPlaceholderReplacer) {
+						matchDetected = true
+					}
+				}
+			}
+		}
+	}
+	return matchDetected
+}
+
+// handleStarPattern handles SELECT * FROM table %%WHERE%% pattern
+func handleStarPattern(queryNodes, patternNodes []sqlparser.SQLNode) bool {
+	patternWhereDetected := false
+	queryWhereDetected := false
+	patternNodeOffset := 0
+	queryNodeOffset := 0
+	for index := 1; index < len(patternNodes); index++ {
+		if index+patternNodeOffset >= len(patternNodes) || index+queryNodeOffset >= len(queryNodes) {
+			return false
+		}
+		if reflect.DeepEqual(patternNodes[index+patternNodeOffset], queryNodes[index+queryNodeOffset]) {
+			continue
+		}
+		if patternSelectExpr, ok := patternNodes[index+patternNodeOffset].(sqlparser.SelectExprs); ok && starFound(patternSelectExpr) {
+			if _, ok := queryNodes[index+queryNodeOffset].(sqlparser.SelectExprs); ok {
+				i := index
+				for i < len(queryNodes) {
```

Comment:
`for ;i<len(queryNodes); i++`

Reply:
Wow, didn't know this loop syntax. Thanks